### PR TITLE
Make timestamps more accurate

### PR DESCRIPTION
**Pull request by martiniil**
## Description

These changes allow the user to get scanner data with reliable timestamps.

* Add scan counter & timestamp to LaserScan API.

Detailed procedure:
Inside the frame with the minimal timestamp (udp package arrival time) we calculate backwards in time for the first ray.

### Things to add, update or check by the maintainers before this PR can be merged.

- [x] #226 
- [x] Public api function documentation
- [x] Architecture documentation reflects actual code structure
- [x] [Tutorials](https://wiki.ros.org/pilz_robots/Tutorials/)
- [x] Overview on [ROS wiki](https://wiki.ros.org/pilz_robots)
- [x] Package Readme ([example pilz_robots](https://github.com/PilzDE/pilz_robots/blob/melodic-devel/README.md))
- [x] Good commit messages ([some tips](https://dev.to/jacobherrington/how-to-write-useful-commit-messages-my-commit-message-template-20n9))
- [x] CHANGELOG.rst updated
- [x] Copyright headers
- [x] Examples

### Review Checklist
- [x] Soft- and hardware architecture (diagrams and description)
- [x] Test review (test plan and individual test cases)
- [x] Documentation describes purpose of file(s) and responsibilities
- [x] Code (coding rules, style guide)

### Release planning (please answer)
- [x] When is the new feature released?
No urgency known, 1.0 release will come.
- [x] Which dependent packages do have to be released simultaneously?
None

### Hardware tests
- [x] Perform hardware tests

### CODE REVIEW

File: standalone/include/psen_scan_v2_standalone/data_conversion_layer/laserscan_conversions.h
Comment (rfeistenauer) [inline standalone/include/psen_scan_v2_standalone/data_conversion_layer/laserscan_conversions.h:152]: Kann man hier durch auslagern in andere Methoden oder notfalls Comments noch bisle Klarheit über den Vorgang herstellen?
get_first_frame()
calculate_time_of_first_ray() <- vorallem das 1000000000/3600 ist denke ich nicht direkt ersichtlich

File: standalone/include/psen_scan_v2_standalone/data_conversion_layer/monitoring_frame_deserialization.h
Comment (rfeistenauer) [inline standalone/include/psen_scan_v2_standalone/data_conversion_layer/monitoring_frame_deserialization.h:20]: Is this necessary?

File: standalone/include/psen_scan_v2_standalone/data_conversion_layer/monitoring_frame_msg.h
Comment (rfeistenauer) [inline standalone/include/psen_scan_v2_standalone/data_conversion_layer/monitoring_frame_msg.h:28]: stamp is a int64 now, so this is probably not necessary anymore

File: standalone/include/psen_scan_v2_standalone/laserscan.h
Comment (rfeistenauer) [inline standalone/include/psen_scan_v2_standalone/laserscan.h:20]: unnecessary?

File: standalone/test/unit_tests/api/unittest_laserscan.cpp
Comment (rfeistenauer) [inline standalone/test/unit_tests/api/unittest_laserscan.cpp:91]: would call it laserScanBuilder or lsBuilder or something that gives the EXPECT something more meaningful than builder.build()

File: standalone/test/unit_tests/api/unittest_laserscan.cpp
Comment (rfeistenauer) [inline standalone/test/unit_tests/api/unittest_laserscan.cpp:16]: is unused?

File: standalone/test/unit_tests/communication_layer/unittest_udp_client.cpp
Comment (rfeistenauer) [inline standalone/test/unit_tests/communication_layer/unittest_udp_client.cpp:17]: unused?

File: test/hw_tests/hwtest_timestamp.py
Comment (rfeistenauer) [inline test/hw_tests/hwtest_timestamp.py:1]: Licence / Copyright notice?

File: test/unit_tests/unittest_laserscan_ros_conversions.cpp
Comment (rfeistenauer) [inline test/unit_tests/unittest_laserscan_ros_conversions.cpp:16]: unused?

**Review by rfeistenauer — CHANGES_REQUESTED** [review]

File: standalone/test/include/psen_scan_v2_standalone/util/integrationtest_helper.h
Comment (rfeistenauer) [inline standalone/test/include/psen_scan_v2_standalone/util/integrationtest_helper.h:147]: rename to make intents of the comparisons clear without knowledge of the tests?

File: standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp
Comment (rfeistenauer) [inline standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp:39]: shouldn't it be `createStampedMsg()`

File: standalone/include/psen_scan_v2_standalone/protocol_layer/scanner_state_machine.h
Comment (martiniil) [inline standalone/include/psen_scan_v2_standalone/protocol_layer/scanner_state_machine.h:241]: We should decide whether we ues `frame` or `msg`, here.

File: standalone/include/psen_scan_v2_standalone/protocol_layer/scanner_state_machine_def.h
Comment (martiniil) [inline standalone/include/psen_scan_v2_standalone/protocol_layer/scanner_state_machine_def.h:148]:
```suggestion
    const data_conversion_layer::monitoring_frame::MessageStamped msg_stamped{ frame, event.timestamp_ };
    informUserAboutTheScanData(msg_stamped);
```

File: standalone/include/psen_scan_v2_standalone/data_conversion_layer/laserscan_conversions.h
Comment (martiniil) [reply standalone/include/psen_scan_v2_standalone/data_conversion_layer/laserscan_conversions.h]: Better now?

File: standalone/test/include/psen_scan_v2_standalone/util/integrationtest_helper.h
Comment (martiniil) [reply standalone/test/include/psen_scan_v2_standalone/util/integrationtest_helper.h:147]: Is it more clear now?

File: standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp
Comment (martiniil) [reply standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp:38]: Maybe, but you already have `Stamped` in the output type. The same holds for `createMsgs()`, `copyMsgWithNewTimestamp()` and `copyMsgWithNewMeasurements()`. I think adding stamped everywhere is a bit too much. What do you think?

Comment (rfeistenauer) [reply standalone/test/unit_tests/data_conversion_layer/unittest_laserscan_conversions.cpp:38]: `const auto frame{ createMsg() };`
you renamed msg with frame and used auto in a lot of tests. If I am new to the project and read the test I would expect frame to be a msg not a stamped one. I think it would get clear pretty fast, but if a `Stmpt` or `Stamped` could save others a few seconds I would change it. The same goes for `createMsgs()`.
Not totaly sure about the other two since the operation is pretty much clear. As a reader you understand what is happening, the only thing here is if you understand what you have to pass in as developer. I'd probably add the stamped here as well.

File: standalone/test/include/psen_scan_v2_standalone/util/integrationtest_helper.h
Comment (rfeistenauer) [inline standalone/test/include/psen_scan_v2_standalone/util/integrationtest_helper.h:148]: Its better already. Maybe we could clearify it even more with those changes?
```suggestion
MATCHER_P2(TimestampInExpectedTimeframe, conversion_time_of_last_scan, prior_scan_timestamp, "")
{
  const int64_t time_since_prior_scan_conversion{ getCurrentTime() - conversion_time_of_last_scan};
  return arg.getTimestamp() > prior_scan_timestamp && arg.getTimestamp() < prior_scan_timestamp + time_since_prior_scan_conversion;
```

File: standalone/include/psen_scan_v2_standalone/protocol_layer/scanner_state_machine.h
Comment (rfeistenauer) [reply standalone/include/psen_scan_v2_standalone/protocol_layer/scanner_state_machine.h:241]: You are right. The idea was stamped ones are `frames` and without its `msg` but thinking about it, it was probably not the best idea.

File: standalone/test/include/psen_scan_v2_standalone/util/integrationtest_helper.h
Comment (rfeistenauer) [reply standalone/test/include/psen_scan_v2_standalone/util/integrationtest_helper.h]: For some reason my suggestion is a separate comment now so see further below for the answer.

File: standalone/include/psen_scan_v2_standalone/protocol_layer/scan_buffer.h
Comment (rfeistenauer) [reply standalone/include/psen_scan_v2_standalone/protocol_layer/scan_buffer.h:90]: there is at least a `d` missing in `stmpd` but yea we can just take your suggestion.

File: standalone/include/psen_scan_v2_standalone/protocol_layer/scanner_state_machine.h
Comment (rfeistenauer) [reply standalone/include/psen_scan_v2_standalone/protocol_layer/scanner_state_machine.h:241]: I remember, why I used frame. Because if you name it msg_stamped you have a lot of msg_stamped.msg_... calls and that just doesn't read as easy as frame.msg_ :/

**Review by rfeistenauer — APPROVED** [review]

File: include/psen_scan_v2/ros_scanner_node.h
Comment (martiniil) [inline include/psen_scan_v2/ros_scanner_node.h:122]: Why is this necessary, suddenly?

File: standalone/include/psen_scan_v2_standalone/laserscan.h
Comment (martiniil) [inline standalone/include/psen_scan_v2_standalone/laserscan.h:36]:
```suggestion
 * - Counter of scan round.
```

File: standalone/test/include/psen_scan_v2_standalone/util/integrationtest_helper.h
Comment (martiniil) [inline standalone/test/include/psen_scan_v2_standalone/util/integrationtest_helper.h:142]: @rfeistenauer What do you think? Should the timestamp also be checked by the equality operator?

Comment (rfeistenauer) [reply standalone/test/include/psen_scan_v2_standalone/util/integrationtest_helper.h:142]: Hmm jea the equality operator does only check measurement related things. 
We should probably test for the rest as well.

Is it intended, that the operator ignores Intensities, timestamp and  scan counter?

Comment (martiniil) [reply standalone/test/include/psen_scan_v2_standalone/util/integrationtest_helper.h:142]: I would move this topic to a separate p.r.

**Comment by martiniil** [comment]
@rfeistenauer From my side this is ready. Can you review the changes to the hardware-tests I did after your approval?

**Pull request merged**
